### PR TITLE
new map method

### PR DESCRIPTION
**Pull request by llogiq**
This adds mapping functions over GenericArrays
Also improves some nits in the doc comments
And ensures panic safety of all operations

### CODE REVIEW

**Comment by llogiq** [comment]
This would fix both #8 and #11, but uses an API that was only stabilized in 1.9.0 (which is current beta).

Alas I see no way to make this work with current stable.

File: src/lib.rs [hunk 1]
```diff
@@ -129,13 +129,53 @@ impl<T, N> DerefMut for GenericArray<T, N> where N: ArrayLength<T> {
...
+        map_inner(&self, f)
+    }
+}
+
+#[inline]
+fn map_inner<S, F, T, N>(list: &[S], f: F) -> GenericArray<T, N>
+where F: Fn(&S) -> T, N: ArrayLength<T> {
+     unsafe {
+        let mut res : GenericArray<T, N> = std::mem::uninitialized();
+        if let Err(e) = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
```
Comment (SSheldon) [inline src/lib.rs]: Should panic catching be used in normal cases? I was under the impression it was only supposed to be used at FFI boundaries.

What if you mapped the uninitialized result in another struct with a `Drop` impl that will only run if the array is fully constructed? Something that might look like (warning, bad pseudo code follows):
``` rust
struct ArrayBuilder {
    res: Option<GenericArray<T, N>>,
    len: usize,
}

impl ArrayBuilder {
    fn new() -> Self {
        ArrayBuilder { res: Some(mem::uninitialized()), len: 0 }
    }

    fn append(item: T) {
        ptr::write(&array[len], item);
        self.len += 1;
    }

    fn finish(self) -> GenericArray {
        self.res.take()
    }
}

impl Drop for ArrayBuilder {
    fn drop(&mut self) {
        if self.len != array.len() {
            mem::forget(self.res.take());
        }
    }
}
```

I think an approach like that could givekn us safety without having to catch panics. But I haven't tried to compile or test any of this code :) Curious to hear what you think!

**Comment by llogiq** [comment]
No, that won't work, because `res` will already have been dropped at that point.

However I'm thinking about a [similar](http://stackoverflow.com/questions/36919842/is-there-a-way-to-pre-un-leak-a-value) approach.

**Comment by SSheldon** [comment]
@llogiq, at what point will res already be dropped? In the drop method? I'm fairly certain a struct's members won't be dropped until after the struct's drop method is called, otherwise every drop would have to be unsafe because you'd be working with memory that was already dropped.

So, if at that point we take res and forget it, I expect that we'll prevent it from being dropped.

**Comment by llogiq** [comment]
Yeah, but as I said, this requires `catch_panic(_)` which will involve some bookkeeping. Using `transmute`s to pre-leak the array will probably be even faster – and possible on stable. Give me a few minutes, I'm currently testing the approach.

**Comment by llogiq** [comment]
There it is. :smile:

**Comment by SSheldon** [comment]
Sorry @llogiq, now I'm just trying to understand :)

> Yeah, but as I said, this requires `catch_panic(_)`

How does writing a struct with our own `Drop` impl that tosses out data that isn't fully initialized require `catch_panic`? We don't need to catch any panics, our `Drop` impl will be called on panic. We can take advantage of that to make sure our struct's state is cleaned up before being dropped, without having to `catch_panic`.

In fact, we could even drop the partially initialized data; we'd know how many elements are valid, so we could `drop_in_place` those before forgetting the whole array.

**Comment by llogiq** [comment]
The problem with that is that dropping the already initialized data calls `.drop(&mut self)` on that data, which could panic itself. Yes, it's very bad form to do so, but as we already deal with a panic, we are in all probability no longer in a valid program state, so a few bytes of leaked memory is an acceptable price to pay.

**Comment by SSheldon** [comment]
> The problem with that is that dropping the already initialized data calls .drop(&mut self) on that data, which could panic itself.

AFAIK panicking while already panicking will result in an abort, so you'd still preserve memory safety in that case :) Having trouble finding a reference on that, but the [Drop docs](https://doc.rust-lang.org/std/ops/trait.Drop.html#tymethod.drop) do say:

> Given that a panic! will call drop() as it unwinds, any panic! in a drop() implementation will likely abort.

**Comment by llogiq** [comment]
Ah, I hadn't seen that. You're right – all it takes is a wrapper that has `Drop`. I may extend this PR or do another one, however you like.

**Comment by SSheldon** [comment]
Heh I've got no say in this codebase, I just ended up here from your blog post because I was curious about different approaches :)

Nice find on `NoDrop`, by the way! I tried out my original approach using `Option` and was foiled by the null-pointer optimization. Always something fun to learn when dealing with unsafe.

**Comment by llogiq** [comment]
@fizyk20 r?

**Comment by fizyk20** [comment]
I had no private computer for a while, so I couldn't take a proper look at this, sorry!

This looks really cool :) If I understand correctly, the latest version uses `NoDrop` so that when there is a panic on initialization, the uninitialized contents of the GenericArray don't get dropped and we are still safe, is that right?

**Comment by llogiq** [comment]
Exactly.

**Comment by fizyk20** [comment]
Awesome :) I'm merging, then, thanks a lot! :)

**Comment by llogiq** [comment]
You're welcome. By the way, I'm still looking out for improvements against the `NoDrop` solution, so if anything comes up, I'll send it to you. :smile:

**Pull request merged**